2022/Q3/artifacts/PoS-tla/Staking.tla

```diff
@@ -113,104 +153,245 @@ Delegate(sender, validator, amount) ==
+
+(*
+* ANTIPATTERN VARIANT
+* CreateSlashes(setSlashes, rate) == LET F(set, slash) == set \union { [slash EXCEPT !.finalRate = rate] }
+*                                    IN ApaFoldSet(F, {}, setSlashes)
+*)
+
+(*
+* At the end of an epoch e:
+* 1. We compute the final slash rates. For each validator with enqueued slashes schedules
+*    to be processed at e, the final rate is the same for each slash.
+* 2. Shift totalDeltas applying any slash as we update the variable.
+* 3. Shift enqueuedSlashes and frozenValidators.
+* 4. Create slashes based on the enqueded ones.
+* 5. Increment epoch.
+*)
+EndOfEpoch ==
+    LET FinalRates == ComputeFinalSlashRates
+    IN
+    /\ totalDeltas' = [ n \in 0..2*UnbondingLength, val \in ValidatorAddrs |-> 
+                        IF n < 2*UnbondingLength
+                        THEN 
+                          IF enqueuedSlashes[0, val] /= {} /\ n >= UnbondingLength + 1
+                          THEN totalDeltas[n+1, val] - AccumulateSlashes(val, FinalRates[val], enqueuedSlashes[0, val])
+                          ELSE totalDeltas[n+1, val]
+                        ELSE 
+                          IF enqueuedSlashes[0, val] /= {}
+                          THEN totalDeltas[n, val] - AccumulateSlashes(val, FinalRates[val], enqueuedSlashes[0, val])
+                          ELSE totalDeltas[n, val]
+                      ]
+    /\ enqueuedSlashes' = [ n \in -1..UnbondingLength, val \in ValidatorAddrs |-> 
+                            IF n < UnbondingLength
+                            THEN enqueuedSlashes[n+1, val]
+                            ELSE {}
+                          ]
+    /\ frozenValidators' = [ n \in 0..UnbondingLength |-> 
+                            IF n < UnbondingLength
+                            THEN frozenValidators[n+1]
+                            ELSE {}
+                          ]
+    /\ slashes' = [ val \in ValidatorAddrs |-> slashes[val] \union CreateSlashes(enqueuedSlashes[0, val], FinalRates[val]) ]
     /\ epoch' = epoch + 1
+    /\ lastTx' = [ id |-> nextTxId, tag |-> "endOfEpoch", fail |-> FALSE,
+                   sender |-> "", toAddr |-> "", value |-> epoch ]
     /\ txCounter' = 0
-    /\ ShiftEpochedVariables  
-    /\ UNCHANGED <<balanceOf, lastTx, nextTxId, failed>>
+    /\ UNCHANGED <<balanceOf, posAccount, bonded, idBonds, unbonded, idUnbonds, idSlashes, nextTxId, failed>>
 
 Next ==
     IF txCounter = TxsEpoch
     THEN
       \* move to the next epoch
-      AdvanceEpoch
+      EndOfEpoch
     ELSE
       \E sender \in UserAddrs:
       \E validator \in ValidatorAddrs:
       \E amount \in Int:
         /\ amount <= MAX_UINT
-        /\ \/ Delegate(sender, validator, amount)
+        \* this guarantees that a validator does not misbehave more than one in an UnbondingLength period.
+        /\ \/ IF validator \notin frozenValidators[0]
+              THEN
```

Comment:
Why requiring this here? Wouldn't it be cleaner to do it inside evidence, just as is done in line 207?

Reply:
Either way works. Maybe I am wrong, but I think this way we do not waste transitions (which are very precious).

Reply:
If you use `NextNoFail` to check, wouldn't that be equivalent in that case? Since no transition can lead to a failed state?

Reply:
Yes, I think so.